### ISSUE TITLE
Speed up node spin-up with placeholders

### ISSUE THREAD

**Issue by choldgraf** [2i2c-org/infrastructure#643]
### Description

We should speed up the time it takes to spin-up our nodes by using placeholders.

Speeding up the nodes would make our hubs perform better whenever there were spikes in activity, or when a user triggers a scale-up event in general. It would help our hubs feel speedier.

### Guide for implementation

:arrow_right: Great info about user placeholders in the [z2jh docs here](https://zero-to-jupyterhub.readthedocs.io/en/latest/administrator/optimization.html#scaling-up-in-time-user-placeholders) and [discourse](https://discourse.jupyter.org/t/planning-placeholders-with-jupyterhub-helm-chart-0-8-tested-on-mybinder-org/213)

**Performance**

To understand whether this would make a bit impact on performance, we could try analyzing the log files from the old U.Toronto hub and compare it with the new one.

**Implementing across cloud providers**

It's unclear whether this would behave the same way across the major cloud providers. Here's what we know about each:

**GKE**: According to [this note in the z2jh docs](https://zero-to-jupyterhub.readthedocs.io/en/latest/administrator/optimization.html#scaling-up-in-time-user-placeholders) this should work on gke at least.

**Azure**:  The original UToronto cluster runs on Azure, and had been using user placeholders :arrow_right: https://github.com/utoronto-2i2c/jupyterhub-deploy/commit/1c7fa04f3e6decfd6b6ebe3c22f613dba03d7d6d (that's not the case anymore since we've migrated it to the pilot-hubs infra as part of https://github.com/2i2c-org/infrastructure/issues/638)

**AWS**: Not sure if/how this works

### Updates and tasks

- [ ] Estimate the performance improvement we'd get with user placeholders
- [ ] Decide if / how we should implement this across the cloud providers

**Comment by GeorgianaElena** [comment]
(*Adding some additional info to this as I believe it doesn't need it's own new issue. Sorry it's a bit unpolished.*)

### Resources
:arrow_right: Great info about user placeholders in the [z2jh docs here](https://zero-to-jupyterhub.readthedocs.io/en/latest/administrator/optimization.html#scaling-up-in-time-user-placeholders) and [discourse](https://discourse.jupyter.org/t/planning-placeholders-with-jupyterhub-helm-chart-0-8-tested-on-mybinder-org/213)

### Context
The original UToronto cluster had been using user placeholders :arrow_right: https://github.com/utoronto-2i2c/jupyterhub-deploy/commit/1c7fa04f3e6decfd6b6ebe3c22f613dba03d7d6d. But that's not the case anymore since we've migrated it to the pilot-hubs infra as part of https://github.com/2i2c-org/infrastructure/issues/638.

### Issue
We need to figure out if we *want or can* re-implement this for the pilot-hubs given that the hubs run on different cloud infra.

Not sure if any analysis of what impact this optimization had on the toronto hub, but maybe the hub log file can help gain more info into this and help make a decision.

### Thoughts
According to this note in [z2jh docs](https://zero-to-jupyterhub.readthedocs.io/en/latest/administrator/optimization.html#scaling-up-in-time-user-placeholders) this should work on gke:

*Further settings may be required for successful use of the pod priority depending on how your cluster autoscaler is configured. This is known to work on GKE, but we don’t know how it works on other cloud providers or kubernetes.*

And since we also have an working example on azure through the UToronto cluster, *maybe* the issue at hand now is to check for aws?

**Comment by choldgraf** [comment]
Thanks @GeorgianaElena for providing this helpful context! I've taken your comment and incorporated it into the top comment, so that we can keep all of the information in one place. I hope that's OK!

I seem to remember a conversation with @yuvipanda and @consideRatio where they said that the user placeholders were not working as well as they thought they were, but I don't know if that was unique to one deployment, or one cloud provider, etc.

**Comment by consideRatio** [comment]
Me and Yuvi have deliberated a lot about this and I'd want to avoid rehashing technical details motivating this suggestion within this issue, but the summarized suggestions are:

- If several user placeholder pods are configured to be used, it is probably better to configure a fewer number of user placeholder pods to run but each having a resource requests matching the least powerful node available for users in the k8s cluster.
- If you simply want to always have at least one node running, configure the node pool to have 1-X in autoscaling range and don't use placeholder pods.

**Comment by damianavila** [comment]
IIRC there were also discussions about node placeholders instead of user placeholders.

**Comment by GeorgianaElena** [comment]
### Update:

UTotoronto folks opened a ticket about cluster scale-up duration that I believe caused them spawn timeouts https://2i2c.freshdesk.com/a/tickets/79. According to the report, this wasn't an isolated event.

Since the utoronto hub is pretty used by users, I believe this will continue to happen. So, I believe we should prioritize this task.

**Comment by consideRatio** [comment]
I added a note about it. I think we have a quite distinct fingerprint of the issue:
1. A user is scheduled to a node, but the node is considered full with "too many pods"
2. Being scheduled to a node - autoscaling the node pool to add new nodes is not done - the scheduler did schedule the pod to a node - it didn't leave it in a pending state in need of a node!

I responded about this in the ticket.
